### PR TITLE
Check scenario before run migration

### PR DESCRIPTION
**Pull request by Settis**
Check if scenario is correct. Can catch following errors:
* duplicating active tasks
* links on inactive tasks or on not presented tasks
* using tasks which not described in tasks.yaml

### CODE REVIEW

**Comment by Wasapl** [comment]
Merged build finished. Test PASSED.

File: cloudferrylib/scheduler/scenario.py [hunk 1]
```diff
@@ -117,3 +123,76 @@ def init_process_migrate(self, path):
...
+                      ", ".join(self.duplicated_tasks))
+            throw_exception = True
+
+        if len(self.missed_tasks):
+            LOG.error("Following tasks not described in tasks.yaml: %s",
+                      ", ".join(self.missed_tasks))
+            throw_exception = True
+
+        if throw_exception:
+            raise Exception("Not valid scenario")
```
Comment (roman-verchikov) [inline cloudferrylib/scheduler/scenario.py]: Let's use some more specific exception, maybe standard `NameError`, or `SyntaxError`

File: cloudferrylib/scheduler/scenario.py [hunk 2]
```diff
@@ -117,3 +123,76 @@ def init_process_migrate(self, path):
...
+            list_value = value[0]
+            if type(list_value) is type(str()):
+                self.check_task(key)
+                self.check_string_list(value)
+                return
+            if type(list_value) is type(dict()):
+                self.check_dict_list(value)
+                return
+            return
+        if type(value) is type(bool()):
```
Comment (roman-verchikov) [inline cloudferrylib/scheduler/scenario.py]: let's use `isinstance` instead of `type() is <something>`

**Comment by Wasapl** [comment]
Merged build finished. Test PASSED.

**Comment by roman-verchikov** [comment]
+1

File: cloudferrylib/scheduler/scenario.py [hunk 3]
```diff
@@ -23,6 +23,9 @@
 import cloudferrylib
 
 from cloudferrylib.base.action import action
+from cloudferrylib.utils import utils as utl
```
Comment (miarmak) [inline cloudferrylib/scheduler/scenario.py]: Roman's favorite) Let's stay with just `utils`.

File: cloudferrylib/scheduler/scenario.py [hunk 4]
```diff
@@ -117,3 +123,84 @@ def init_process_migrate(self, path):
...
+        self.check_dict_list(self.scenario)
+
+        # check missed links
+        for link in self.links:
+            if link not in self.active_tasks:
+                self.missed_links.add(link)
+
+        throw_exception = False
+        # log all errors if needed
+        if len(self.missed_links):
```
Comment (miarmak) [inline cloudferrylib/scheduler/scenario.py]: Why do we need `len` function here? I guess `if self.missed_links:` will have the same effect.

File: cloudferrylib/scheduler/scenario.py [hunk 5]
```diff
@@ -117,3 +123,84 @@ def init_process_migrate(self, path):
...
+            LOG.error("Scenario has missed links: %s",
+                      ", ".join(self.missed_links))
+            throw_exception = True
+
+        if len(self.duplicated_tasks):
```
Comment (miarmak) [inline cloudferrylib/scheduler/scenario.py]: The same as above

File: cloudferrylib/scheduler/scenario.py [hunk 6]
```diff
@@ -117,3 +123,84 @@ def init_process_migrate(self, path):
...
+            LOG.error("Scenario has duplicated tasks: %s",
+                      ", ".join(self.duplicated_tasks))
+            throw_exception = True
+
+        if len(self.missed_tasks):
```
Comment (miarmak) [inline cloudferrylib/scheduler/scenario.py]: And here

File: cloudferrylib/scheduler/scenario.py [hunk 7]
```diff
@@ -117,3 +123,84 @@ def init_process_migrate(self, path):
...
+        if throw_exception:
+            raise ScenarioError(self)
+
+    def check_string_list(self, node_list):
+        for node in node_list:
+            self.links.add(node)
+
+    def check_node(self, node):
+        key, value = node
+        if type(value) is type(list()):
```
Comment (miarmak) [inline cloudferrylib/scheduler/scenario.py]: Maybe it is better to have `if isinstance(value, list):` instead of types compare?

**Comment by roman-verchikov** [comment]
Please fix inlines

**Comment by Settis** [comment]
Done.

**Comment by Wasapl** [comment]
Merged build finished. Test PASSED.

**Comment by roman-verchikov** [comment]
+1

**Comment by miarmak** [comment]
Looks good, +1. Thank!

**Comment by roman-verchikov** [comment]
Got two +1s, can be merged

**Pull request merged**